core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -0,0 +1,159 @@
+  );
+
+  private static final Accessor<StructLike> FILENAME_ACCESSOR = POSITION_DELETE_SCHEMA.accessorForField(1);
+  private static final Accessor<StructLike> POSITION_ACCESSOR = POSITION_DELETE_SCHEMA.accessorForField(2);
+
+  private Deletes() {
+  }
+
+  public static <T> CloseableIterable<T> positionFilter(CloseableIterable<T> rows, Function<T, Long> rowToPosition,
+                                                        CloseableIterable<Long> posDeletes) {
+    return new PositionDeleteFilter<>(rows, rowToPosition, posDeletes);
+  }
+
+  public static CloseableIterable<Long> deletePositions(String dataLocation, CloseableIterable<StructLike> posDeletes) {
+    return deletePositions(dataLocation, ImmutableList.of(posDeletes));
+  }
+
+  public static CloseableIterable<Long> deletePositions(String dataLocation,
+                                                        List<CloseableIterable<StructLike>> deleteFiles) {
+    DataFileFilter locationFilter = new DataFileFilter(dataLocation);
+    List<CloseableIterable<Long>> positions = Lists.transform(deleteFiles, deletes ->
+        CloseableIterable.transform(locationFilter.filter(deletes), row -> (Long) POSITION_ACCESSOR.get(row)));
+
+    return new SortedMerge<>(Long::compare, positions);
+  }
+
+  private static class PositionDeleteFilter<T> extends CloseableGroup implements CloseableIterable<T> {
+    private final CloseableIterable<T> rows;
+    private final Function<T, Long> extractPos;
+    private final CloseableIterable<Long> deletePositions;
+
+    private PositionDeleteFilter(CloseableIterable<T> rows, Function<T, Long> extractPos,
+                                 CloseableIterable<Long> deletePositions) {
+      this.rows = rows;
+      this.extractPos = extractPos;
+      this.deletePositions = deletePositions;
+    }
+
+    @Override
+    public CloseableIterator<T> iterator() {
+      CloseableIterator<Long> deletePosIterator = deletePositions.iterator();
+
+      CloseableIterator<T> iter;
+      if (deletePosIterator.hasNext()) {
+        iter = new PositionFilterIterator(rows.iterator(), deletePosIterator);
+      } else {
+        iter = rows.iterator();
+        try {
+          deletePosIterator.close();
+        } catch (IOException e) {
+          throw new UncheckedIOException("Failed to close delete positions iterator", e);
+        }
+      }
+
+      addCloseable(iter);
+
+      return iter;
+    }
+
+    private class PositionFilterIterator extends FilterIterator<T> {
+      private final CloseableIterator<Long> deletePosIterator;
+      private long nextDeletePos;
+
+      protected PositionFilterIterator(CloseableIterator<T> items, CloseableIterator<Long> deletePositions) {
+        super(items);
+        this.deletePosIterator = deletePositions;
+        this.nextDeletePos = deletePosIterator.next();
+      }
+
+      @Override
+      protected boolean shouldKeep(T row) {
+        long currentPos = extractPos.apply(row);
```

Comment:
Does this assume rows are ordered by position?

Reply:
Yes, it assumes that positions are ascending, like the delete positions.